datafusion/expr/src/expr_fn.rs

```diff
@@ -676,6 +679,266 @@ pub fn interval_month_day_nano_lit(value: &str) -> Expr {
     Expr::Literal(ScalarValue::IntervalMonthDayNano(interval))
 }
 
+/// Extensions for configuring [`Expr::AggregateFunction`]
+///
+/// Adds methods to [`Expr`] that make it easy to set optional aggregate options
+/// such as `ORDER BY`, `FILTER` and `DISTINCT`
+///
+/// # Example
+/// ```no_run
+/// # use datafusion_common::Result;
+/// # use datafusion_expr::{AggregateUDF, col, Expr, lit};
+/// # use sqlparser::ast::NullTreatment;
+/// # fn count(arg: Expr) -> Expr { todo!{} }
+/// # fn first_value(arg: Expr) -> Expr { todo!{} }
+/// # fn main() -> Result<()> {
+/// use datafusion_expr::ExprFunctionExt;
+///
+/// // Create COUNT(x FILTER y > 5)
+/// let agg = count(col("x"))
+///    .filter(col("y").gt(lit(5)))
+///    .build()?;
+///  // Create FIRST_VALUE(x ORDER BY y IGNORE NULLS)
+/// let sort_expr = col("y").sort(true, true);
+/// let agg = first_value(col("x"))
+///   .order_by(vec![sort_expr])
+///   .null_treatment(NullTreatment::IgnoreNulls)
+///   .build()?;
+/// # Ok(())
+/// # }
+/// ```
+pub trait ExprFunctionExt {
```

Comment:
Rather than trying to unify treatment of aggregate and window functions (which are almost, but not quite the same), it might be worth considering creating a `WindowExt` that mirrors closely the `AggregateExt`

While they would likely have substantial duplication I think the code might be simpler to reason about and use

Also, it would not be an API change, which is an added bonus

Reply:
The one problem I see there is that we would end up with Expr implementing two traits with identical function calls, which would be less user friendly.

```
pub trait AggregateExt {
    fn order_by(self, order_by: Vec<Expr>) -> AggregateBuilder;
    ...
}

pub trait WindowExt {
    fn order_by(self, order_by: Vec<Expr>) -> WindowBuilder;
    ...
}

impl AggregateExt for Expr {
    fn order_by(self, order_by: Vec<Expr>) -> AggregateBuilder {
        ...
    }
}

impl WindowExt for Expr {
    fn order_by(self, order_by: Vec<Expr>) -> WindowBuilder {
        ...
    }
}
```

Then when the user goes to create their aggregate or window I would expect an error on having multiple methods in scope or the user would have to specify which trait they're working with. Right?

On a slightly separate topic, One thing I was trying to figure out this morning was if we need a builder structure. Would it be simpler to have each of these return an Expr? I can see one issue in that we would want to generate some errors if the user called one of these functions on an Expr that is inappropriate. I was thinking we could do something along the lines of

```
impl Expr {
    pub fn order_by(mut self, order_by: impl Into<Option<Vec<Expr>>>) -> Result<Expr> {
        match &mut self {
            Expr::AggregateFunction(udaf) => {
                udaf.order_by = order_by.into();
            }
            Expr::WindowFunction(udwf) => {
                udwf.order_by = order_by.into()
            }
            _ => {
                return plan_err!("Unable to set `order_by` on an Expression that is not an AggregateFunction or WindowFunction.");
            }
        }
        Ok(self)
    }
    ...
}
```

Then the user can do something like

```
let agg = first_value(col("x")).order_by(vec![sort_expr])?
```

But maybe that's less ergonomic than keeping with the builder structure.

Reply:
Here is an alternate approach: https://github.com/apache/datafusion/pull/11582

In that PR instead of using a builder structure, each of these calls like `partition_by(expr)` instead returns a `Result<Expr>`.

I think this makes it pretty ergonomic, but it is an API change. 

From one of the unit tests we have the old method:

```
async fn test_aggregate_ext_filter() {
    let agg = first_value_udaf()
        .call(vec![col("i")])
        .order_by(vec![col("i").sort(true, true)])
        .filter(col("i").is_not_null())
        .build()
        .unwrap()
        .alias("val");
```

And the new method:
```
async fn test_aggregate_ext_filter() -> Result<()> {
    let agg = first_value_udaf()
        .call(vec![col("i")])
        .order_by(vec![col("i").sort(true, true)])?
        .filter(col("i").is_not_null())?
        .alias("val");
```

*Maybe* it doesn't make a lot of difference, which is why I put it on a different branch. If we want to stick with the builder method, then this PR is ready to go assuming we come to agreement on if this should be a single trait or broken into two for Aggregate vs Window.

Reply:
> The one problem I see there is that we would end up with Expr implementing two traits with identical function calls, which would be less user friendly.

I agree that is not a good user experience

> then this PR is ready to go assuming we come to agreement on if this should be a single trait or broken into two for Aggregate vs Window.

I wonder if we could do something like this:

```rust
trait WindowExt: AggregateExt {
...
}
```

which is kind of like Rust's inheritance -- it basically says that anything that implements `WindowExt` also implements `AggregateExt`

And then the builder could be extended for window specific functions (and perhaps `build` made to error if the underlying aggregate function couldn't be used as a window function)